### DISCUSSION TITLE
A way to access screen normals in canvas_item shaders

### DISCUSSION THREAD

**Discussion by demolen** [godotengine/godot-proposals#10127] [2D]
I have made a 2D GI Post processing effect. To use the normals, Currently the way I deal with it is to copy all the Sprites with CanvasTexture to another subViewport, render their normals and sample this subViewport in my GI shader. I would prefer a more straight forward solution as the current way of doing this is not really a "solution".

Before opening a proposal for this I would like to hear about why this is not already implemented, what are the alternatives and what would be a possible implementation scenario.

**Comment by demolen** [comment]
I start with the possible solutions:
1- use the r channel to choose a color off a lookup table would leave channels g and b free to be used for normals and specular combined.

**Comment by Jesusemora** [comment]
> Before opening a proposal for this I would like to hear about why this is not already implemented, what are the alternatives and what would be a possible implementation scenario.

Godot uses the forward pipeline. What you are asking is commonly found with the deferred pipeline.
Keeping a buffer for normals is just not necesary with forward and would be wasted memory only used for this situation.
I could be wrong though.
